### PR TITLE
Add typed events for the filesystem/scanner

### PR DESCRIPTION
**Pull request by ChristophWurst**
Continuation of #18348 and #18350.

For https://docs.nextcloud.com/server/stable/developer_manual/app/hooks.html#filesystem-scanner

 - [x] scanFile (string $absolutePath)
 - [x] scanFolder (string $absolutePath)
 - [x] postScanFile (string $absolutePath)
 - [x] postScanFolder (string $absolutePath)

File scanner and app dir scanner still work fine here. Let's see what CI thinks.

### CODE REVIEW

File: apps/files/lib/BackgroundJob/ScanFiles.php [hunk 1]
```diff
@@ -41,39 +41,31 @@ class ScanFiles extends \OC\BackgroundJob\TimedJob {
...
 	/** @var ILogger */
 	private $logger;
+
 	/** Amount of users that should get scanned per execution */
 	const USERS_PER_SESSION = 500;
 
 	/**
 	 * @param IConfig|null $config
 	 * @param IUserManager|null $userManager
-	 * @param IDBConnection|null $dbConnection
```
Comment (ChristophWurst) [inline apps/files/lib/BackgroundJob/ScanFiles.php]: @icewind1991 please have a look at this. The `$dbConnection` was never assigned, hence always `null`. Yet we pass it to the scanner.

File: lib/public/Files/Events/FolderScannedEvent.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+	/** @var string */
+	private $absolutePath;
+
+	/**
+	 * @param string $absolutePath
+	 *
+	 * @since 18.0.0
+	 */
+	public function __construct(string $absolutePath) {
```
Comment (rullzer) [inline lib/public/Files/Events/FolderScannedEvent.php]: These events are all rather similar...

For a followup we could merge them and just have them extend absract events? That way you could listen to a generatic BeforeScannedEvent?

**Review by rullzer — APPROVED** [review]
But you are missing the composer update ;)

**Review by icewind1991 — APPROVED** [review]

File: apps/files/lib/BackgroundJob/ScanFiles.php (hunk 1 above)
Comment (icewind1991) [reply apps/files/lib/BackgroundJob/ScanFiles.php]: makes sense

**Comment by rullzer** [comment]
Tests do :boom:

**Comment by MorrisJobke** [comment]
ref #14552

**Pull request merged**